x-pack/plugin/mapper-aggregate-metric/src/main/java/org/elasticsearch/xpack/aggregatemetric/aggregations/metrics/AggregateMetricBackedMaxAggregator.java

```diff
@@ -0,0 +1,200 @@
+ * you may not use this file except in compliance with the Elastic License.
+ */
+package org.elasticsearch.xpack.aggregatemetric.aggregations.metrics;
+
+import org.apache.lucene.index.LeafReader;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.PointValues;
+import org.apache.lucene.search.CollectionTerminatedException;
+import org.apache.lucene.search.ScoreMode;
+import org.apache.lucene.util.Bits;
+import org.apache.lucene.util.FutureArrays;
+import org.elasticsearch.common.lease.Releasables;
+import org.elasticsearch.common.util.BigArrays;
+import org.elasticsearch.common.util.DoubleArray;
+import org.elasticsearch.index.fielddata.NumericDoubleValues;
+import org.elasticsearch.index.fielddata.SortedNumericDoubleValues;
+import org.elasticsearch.search.DocValueFormat;
+import org.elasticsearch.search.MultiValueMode;
+import org.elasticsearch.search.aggregations.Aggregator;
+import org.elasticsearch.search.aggregations.InternalAggregation;
+import org.elasticsearch.search.aggregations.LeafBucketCollector;
+import org.elasticsearch.search.aggregations.LeafBucketCollectorBase;
+import org.elasticsearch.search.aggregations.metrics.InternalMax;
+import org.elasticsearch.search.aggregations.metrics.NumericMetricsAggregator;
+import org.elasticsearch.search.aggregations.pipeline.PipelineAggregator;
+import org.elasticsearch.search.aggregations.support.ValuesSourceConfig;
+import org.elasticsearch.search.internal.SearchContext;
+import org.elasticsearch.xpack.aggregatemetric.aggregations.support.AggregateMetricsValuesSource;
+import org.elasticsearch.xpack.aggregatemetric.mapper.AggregateDoubleMetricFieldMapper.Metric;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+
+import static org.elasticsearch.search.aggregations.metrics.MinAggregator.getPointReaderOrNull;
+
+class AggregateMetricBackedMaxAggregator extends NumericMetricsAggregator.SingleValue {
+
+    private final AggregateMetricsValuesSource.AggregateDoubleMetric valuesSource;
+    final DocValueFormat formatter;
+
+    final String pointField;
+    final Function<byte[], Number> pointConverter;
+
+    DoubleArray maxes;
+
+    AggregateMetricBackedMaxAggregator(
+        String name,
+        ValuesSourceConfig config,
+        AggregateMetricsValuesSource.AggregateDoubleMetric valuesSource,
+        SearchContext context,
+        Aggregator parent,
+        List<PipelineAggregator> pipelineAggregators,
+        Map<String, Object> metaData
+    )
+        throws IOException {
+        super(name, context, parent, pipelineAggregators, metaData);
+        this.valuesSource = valuesSource;
+        if (valuesSource != null) {
+            maxes = context.bigArrays().newDoubleArray(1, false);
+            maxes.fill(0, maxes.size(), Double.NEGATIVE_INFINITY);
+        }
+        this.formatter = config.format();
+        this.pointConverter = getPointReaderOrNull(context, parent, config);
```

Comment:
I think this won't end up working, or rather, will never return a valid pointConverter and so all the associated BKD logic is not needed.  E.g. `MinAggregator#getPointReaderOrNull()` fetches the field context and then does naive `instanceof` checks to see what kind of field it is (number or date atm).  Which I believe won't match because the field here is actually the pre-aggregated field class.

We'd need some way to tell that function (or create a new one) which metric to get out of the pre-agg field, which is harder since it's in Core.

It's a nice optimization but also pretty niche... I think it would be ok to remove all the point reader / bkd optimization from these aggregators to keep them simple.  We can add them back later.

Reply:
Removed all point reader / bkd optimization